src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
> @jdom: This is not compatible with supporting a different IGrainActivator

> @jason-bragg: "Yes it does. Grain activators simply can't select constructors differently for non-application grains. If we want to specifically make constructor selection part of the grain activator, we can do that, but that can only be supported for application grains, which is of little benefit compared to the nuance and complexity it would introduce. It can also be added at later time if/when needed."

We have to construct app and system grains the *same* way IMO, since we support user specified container usage and user specified grain activators.

The problem here is that we like to have some information (the ctor selected by the configured DI container) which is not exposed through the DI abstraction since we don't have access to it, since the instance construction is internal to every DI container.

I like to have attributes on the constructor arguments, but since DI abstraction does not have any way to support this kind of metadata whatever we are adding here belongs to DI land. Maybe using attributes (however it is nice) the most unfortunate thing we can support in the initial implementation, due the lack of support.

What we can require from app developers that if they like to use facets, then go though some well defined registration methods if they're using facets, so during registration we can have ways to extract the metadata we need for our default implementation.

My 2 cents against ```GrainTypeData``` use of reflection is that we're trying to drive an effort where all of the reflection/grain metadata would be stored, accessed and populated in one place and adding such logic into classes going against this effort.

Since DI is deeply involved I'd like to see tests running the same code with Autofac for example to make us confident that other conforming containers can work nicely without providing extra code from the app developer who is using a 3rd party container.